app/src/main/java/net/squanchy/schedule/filterschedule/FilterSchedule.kt

```diff
@@ -0,0 +1,135 @@
+import android.support.v7.widget.LinearLayoutManager
+import android.support.v7.widget.RecyclerView
+import android.view.LayoutInflater
+import android.view.ViewGroup
+import android.widget.CheckBox
+import io.reactivex.Observable
+import io.reactivex.android.schedulers.AndroidSchedulers
+import io.reactivex.disposables.Disposable
+import io.reactivex.functions.BiFunction
+import io.reactivex.schedulers.Schedulers
+import kotlinx.android.synthetic.main.activity_track_filters.*
+import net.squanchy.R
+import net.squanchy.schedule.domain.view.Track
+import net.squanchy.service.repository.TracksRepository
+
+class FilterScheduleActivity : AppCompatActivity() {
+
+    private lateinit var tracksRepository: TracksRepository
+    private lateinit var tracksFilter: TracksFilter
+    private lateinit var trackAdapter: TrackFiltersAdapter
+
+    private var subscription: Disposable? = null
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        setContentView(R.layout.activity_track_filters)
+
+        val component = tracksFilterComponent(this)
+        tracksRepository = component.tracksRepository()
+        tracksFilter = component.tracksFilter()
+
+        trackAdapter = TrackFiltersAdapter(this)
+
+        trackFiltersList.layoutManager = LinearLayoutManager(this)
+        trackFiltersList.adapter = trackAdapter
+    }
+
+    override fun onStart() {
+        super.onStart()
+
+        subscription = Observable.combineLatest(tracksRepository.tracks(), tracksFilter.selectedTracks, combineIntoCheckableTracks())
+            .subscribeOn(Schedulers.computation())
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe {
+                trackAdapter.updateTracks(it) { track, selected ->
+                    val selectedTracks = it.filter { it.second }.map { it.first }.toSet()
+                    val newSelectedTracks = if (selected) selectedTracks + track else selectedTracks - track
+                    tracksFilter.updateSelectedTracks(newSelectedTracks)
+                    trackAdapter.notifyItemChanged(it.map { checkableTrack -> checkableTrack.first }.indexOf(track))
+                }
+            }
+    }
+
+    private fun combineIntoCheckableTracks(): BiFunction<List<Track>, Set<Track>, List<CheckableTrack>> {
+        return BiFunction { tracks, selectedTracks ->
+            tracks.map { CheckableTrack(it, selectedTracks.contains(it)) }
+        }
+    }
+
+    override fun onStop() {
+        super.onStop()
+
+        subscription?.dispose()
+    }
+}
+
+private class TrackFiltersAdapter(context: Context) : RecyclerView.Adapter<TrackViewHolder>() {
+
+    init {
+        setHasStableIds(true)
+    }
+
+    private val layoutInflater = LayoutInflater.from(context)
+
+    private var checkableTracks: List<CheckableTrack> = emptyList()
+    private var trackStateChangeListener: OnTrackSelectedChangeListener? = null
```

Comment:
I'd have this `lateinit` rather than nullable, the rationale behind that is that if we fail to set it, the app should crash (instead of falling in a non-functional state that the user doesn't know how to act upon)

Reply:
Yes! You're right. 7e90b3bdaaf5c6d2e5261300c4c25a7dfd447c5c